### DISCUSSION TITLE
How to detect all Siemens S7 PLCs in the network?

### DISCUSSION THREAD

**Discussion by mateusz-glowinski** [gijzelaerr/python-snap7#402] [Q&A] [answered]
Hi,
is there a way to detect all Siemens S7 PLCs in the network using this package?

**Comment by Cube707** [comment]
I have just started using this package and don't know if its posible to do this using snap7.

Howerver, an alternative would be the [`SIMATIC Automation Tool`](https://support.industry.siemens.com/cs/ww/de/view/98161300). You can use  the provided GUI or the C# API from code. It requires a Licence though...

**Reply by mateusz-glowinski** [reply]
Thank you @Cube707 for a tip. Indeed this required license, that is why I though about `snap7` :)

**Reply by swamper123** [reply]
Hey ho, better late then never. ;)

The answer is: yesn't. ^^

No - because snap7 hasn't got any "Broadcast" command and we haven't got a feature for that. I am not sure, but I think the device would need to support snap7 with all it's needed configs.

Yes - if you brutforce things (at least I could imagine that). IP-Addresses may not be the problem, but rack & slot perhaps. But even then, the sps may need to support snap7 with all it's needed config.

**Reply by mateusz-glowinski** [reply]
Sorry, it's not clear to me.
What do you mean by ` brutforce things`?

**Reply by Cube707** [reply]
What he means is that you can try to connect to every possible IP adress and the one that work are from connected devices...

**Reply by mateusz-glowinski** [reply]
oh, clear :) thank you for a tip

**Comment by gijzelaerr** [comment, marked as answer]
Closing as answered. Network device discovery is outside the scope of this library.

**Reply by Cube707** [reply]
I might add that TIA portal uses DCP protocol for network discovery and only after that uses S7 protocol to comunicate.

So for anybody stumebling accross this in the future, the better path might be to use [`pnio-dcp`](https://pypi.org/project/pnio-dcp/) package and its `identify_all()` function to identify the devices and than open S7 connections where applicable

**Reply by gijzelaerr** [reply]
Thanks for adding this information! I've put adding DCP support to the roadmap https://github.com/gijzelaerr/python-snap7/issues/622

**Reply by gijzelaerr** [reply]
and implemented in a PR https://github.com/gijzelaerr/python-snap7/pull/634
